synapse/lib/stormlib/model.py

```diff
@@ -812,3 +814,189 @@ async def _methCopyTags(self, src, dst, overwrite=False):
+                      {'name': 'prefer_v22', 'type': 'bool', 'default': False,
+                       'desc': '''
+                        Try to renormalize using the :v2_2 prop instead of the
+                        primary property. This can be especially useful when the
+                        primary property is a valid but incorrect CPE string.
+                        '''},
+                      {'name': 'force', 'type': 'bool', 'default': False,
+                       'desc': 'Perform fixups even if the primary property and :v2_2 are valid.'},
+                  ),
+                  'returns': {'type': 'boolean', 'desc': 'Boolean indicating if the migration was successful.'}}},
+    )
+    _storm_lib_path = ('model', 'migration', 's')
+
+    def getObjLocals(self):
+        return {
+            'itSecCpeFixup': self._itSecCpeFixup,
+        }
+
+    async def _itSecCpeFixup(self, n, prefer_v22=False, force=False):
+
+        if not isinstance(n, s_node.Node):
+            raise s_exc.BadArg(mesg='$lib.model.migration.s.itSecCpeFixup() argument must be a node.')
+
+        if n.form.name != 'it:sec:cpe':
+            raise s_exc.BadArg(f'itSecCpeFix only accepts it:sec:cpe nodes, not {n.form.name}')
+
+        prefer_v22 = await s_stormtypes.tobool(prefer_v22)
+        force = await s_stormtypes.tobool(force)
+
+        layr = self.runt.snap.wlyr
+        self.runt.confirm(('node', 'prop', 'set', 'it:sec:cpe'), gateiden=layr.iden)
+        self.runt.confirm(('node', 'data', 'set', 'migration.s.itSecCpeFixup'), gateiden=layr.iden)
+
+        curv = n.repr()
+        reprvalu = f'it:sec:cpe={curv}'
+
+        nodedata = await n.getData('migration.s.itSecCpeFixup', {})
+        if nodedata.get('status') == 'success' and not force:
+            if self.runt.debug:
+                mesg = f'DBG: itSecCpeFixup({reprvalu}): Node already migrated.'
+                await self.runt.printf(mesg)
+            return True
+
+        modl = self.runt.model.type('it:sec:cpe')
+
+        valu23 = None
+        valu22 = None
+        invalid = ''
+
+        # Check the primary property for validity.
+        cpe23 = s_infotech.cpe23_regex.match(curv)
+        if cpe23 is not None and cpe23.group() == curv:
+            valu23 = curv
+
+        # Check the v2_2 property for validity.
+        v2_2 = n.props.get('v2_2')
+        if v2_2 is not None:
+            rgx = s_infotech.cpe22_regex.match(v2_2)
+            if rgx is not None and rgx.group() == v2_2:
+                valu22 = v2_2
+
+        async with self.runt.snap.getNodeEditor(n) as proto:
+
+            # If both values are populated, this node is valid
+            if valu23 is not None and valu22 is not None and not force:
+                if self.runt.debug:
+                    mesg = f'DBG: itSecCpeFixup({reprvalu}): Node is valid, no migration necessary.'
+                    await self.runt.printf(mesg)
+
+                await proto.setData('migration.s.itSecCpeFixup', {
```

Comment:
Maybe this was already discussed, but adding this on success seems like un-needed writes.  Could have this function return retn style (ok, {<info>})?  If nodedata is only written on failure then it would be easy to lift all of them by nodedata key.

Reply:
There were two reasons for adding this to nodedata:
- We can test for status==success in this function (it's above here) and exit early if the node has already been migrated
- If the node is successfully migrated later in this function, we set status=success and a list of properties that were updated

I'll bring it up at scrum later to see what the group thinks.